README.md
# COMPE470L UART Lab

<details>
	<summary> The Instructions </summary>


[Professor Ken Arnold's instruction video](https://drive.google.com/file/d/1Q-ztf6lWboTvkhMa8we5UHa06pon7dVH/view)

This assignment is to create two state machine designs in Verilog and demonstrate them on the FPGA board:

1. **Tx:** The simpler of the two. When an 8 bit value is loaded into a register using the 8 DIP switches for the number and a push button for the "load" signal, it shifts the byte out in asynchronous serial format (initially at 9600 bits per second, later at an arbitrary, programmable data rate).  That begins with a start bit (0), followed by the 8 data bits LSB first, and a stop (1) bit.

2. Rx: The tough one, receiving a byte in the format above and displaying it using the LEDs. Your Rx will have to detect the start bit, with 1/2 bit period to confirm a valid start bit, then sample in the middle of each bit interval shifting each bit into an 8 bit register that drives the 8 LEDs on the IO board.

<details>
	<summary> **Part 1 - Transmit** </summary>
For this specific assignment, in part 1 you must implement a UART that takes paralllel input data from the switches and buttons, and produces a serial output on one of the FPGA output pins.  In order to do that, you must also create a clock with an appropriate frequency to operate the UART from the on-board oscillator connected to the FPGA.  The clock frequency should be higher than the data rate to allow for the requirements for part 2 below, most UARTs use a clock that is 16x the data rate. Capture the serial output data on the scope or logic analyzer abd confirm the serial output data is correct and that the bit period is 1/9600th second long.
	</details>

<details>
	<summary> Part 2 - Receive </summary>
In part 2, you will design a serial to parallel receiver that will receive the asynchronous data from your transmitter in part 1 above, and convert it into an 8 bit parallel word for display on the LEDs on the I/O board.
	</details>

Ultimately, you will be implementing the core subset of transmit/receive functions of a device similar to the SCC2691 serial UART chip in the file listed below, _(Note: no such file was found in the professor's lab instructions.)_ so you should review the transmit buffer empty and receive buffer full status registers of that device.  For full credit, your final UART design should implement the receive buffer full and transmit buffer empty bits. You will need to take the raw FPGA input clock ~~(8MHz for the older version of the board, 50MHz for the newer rev A board that has DIYchips.com written in white letters on the top of the board)~~ _(This instruction was originally written by the teacher with the [FPGADEVS6 board from DIYchips.com](http://diychips.com/fpgadevs6.html) in mind. This project has been implemented instead on the [Digilent Basys 3](https://reference.digilentinc.com/reference/programmable-logic/basys-3/start) which has a single 100 MHz oscillator.)_ and convert that clock into an appropriate clock for your two state machines.
</details>

<details>
	<summary> The Logic </summary>

![The Basic Logic](https://github.com/gameguy95/UART_TX/raw/master/docs/UART%20Tx%20Logic%20Simplified.png "The Basic Logic")

This is a basic flowchart of the logic I used in the design. If you want to understand the full picture, look at the code, or at the picture `UART_TX/docs/UART Tx Logic.png`
      

</details>

<details>
	<summary> The Lab Report </summary>


<details>
	<summary>What Did I Do To Complete the Lab?</summary>
I watched the video on blackboard about UART Tx and then drew out a first version of my state machine logic, as shown in the Logic section above.

Then I started trying to write out a basic state machine code using this. This didn’t really work well, though, and I had to make a lot of changes to get it to even output anything. After all those changes, I got it to output garbage, which was better than nothing. Every time a problem came up, I google searched what might be causing that, got a new problem, google searched, etc.
	</details>

<details>
	<summary>What Challenges Did I Have With This Lab?</summary>
+ I was getting no data output at first, because the board didn’t like the way that I was trying to read the data and then concatenate it with the start and stop bits into a single register titled _transmit._
+ Laptop BSOD’d from trying to synthesize my code. Apparently using Vivado’s internal debugging tools sometimes causes problems for… reasons. Still not sure why they suddenly caused problems when they worked fine for hours before that.
+ Swapped back and forth between for loops and if loops several times, because I was having trouble getting if loops to output anything and I thought that outputting something that seemed like garbage was probably better than outputting nothing. Turns out my if loops were just set up wrong, because I had to use if loops in my final version.
	</details>

<details>
	<summary>What Did I Learn From This Lab?</summary>
* For loops in Verilog
    * Are synthesized by essentially copy/pasting the for loop over and over for every value of , so my 2 nested for loops were essentially making 8*10415 = 83320 statements.
    * do not execute sequentially, so my code was running 83320 iterations of the loop and only outputting the final result
	</details>

</details>


Changelog.txt
Revision 0.19
	- Trying nested if statements again, because a google search said that for loops in verilog aren't sequential, meaning the entire for loop was operating in one clock cycle. That explains why the highest bit was the only one that made a visible change in the output waveform. 
Revision 0.18
	- Removed fancy idle_bit, start_bit, and stop_bit parameters because they were unneccesary and I feel like they might be part of why the synthesis was taking 10+ minutes 
	- Initialized data, bit, and counter registers to 0 because... can't hurt, right? 
Revision 0.17
	- Tried changing all the blocking = signs inside the always block to nonblocking <= signs
Revision 0.16
	- replaced 	current_state	 and 	next_state	 with a single state variable 	state	because Vivado says it merges the two anyways. 
	- removed default state from case statement because Vivado says default case is never used
	- removed unneccesary inputs IO_SWITCH 8 to 15 that were never used. 
Revision 0.15
	- Found an error where I typed	 for (bit = 0; counter < 7; bit = bit + 1)	when it should have been	for (bit = 0; bit < 7; bit = bit + 1)
Revision 0.14 
	- Solved error that caused synthesis of Revision 0.13 caused my laptop to BSOD twice. Solved by removing debug declarations from XDC file. Not sure why they cause a proble because they worked for the last several revisions. 
	- Restored for loop logic because I wasn't getting any output with my if loop logic. Now it at least outputs some incorrect stuff, instead of always asserting 0 when button pushed.	
 Revision 0.13 
	- Adding more states to try to solve the problem of no data output. After using LEDs for Debugging in Revision 0.12, it looks like the input data is being read from the switches correctly, because using	 assign IO_LED = data;	 turns on the correct LEDs. But trying to concatenate 	transmission = {stop_bit, data, start_bit};	 and outputting that to the LED's doesn't turn on any LEDs.  
 Revision 0.12 
	- Still no data output. Added LEDs for debugging.  
 Revision 0.11 
	- 0.10 still wouldn't output anything. I think I tried to get too fancy by setting data size to [number_data_bits-1:0] instead of hardcoding size
 Revision 0.10 
	- Fixed logic error on line 120 where I put a 1 when it should be a 0, resulting in the Tx pin always asserting 0 after button pushed
 Revision 0.09 
	- Apparently I never incremented counter in revision 0.08. Fixed that. 
 Revision 0.08 
	- I think the for loop logic is messing me up. Trying if statements instead. 
 Revision 0.07 
	- tried changing all nonblocking statements to blocking statements
 Revision 0.06 
	- changed read logic from      if (IO_BTN_C ) begin   to     if (IO_BTN_C && ~transmit ) begin  
 Revision 0.05 
	- Removed read state, moved into an if statement 
 Revision 0.04 
	- Added baud rate generation 
 Revision 0.03 
	- Move transmit logic into for loop
 Revision 0.02 
	- First attempts at basic code
 Revision 0.01 
	- File Created

UART_TX.srcs/sources_1/new/UART_TX.v
`timescale 1ns / 1ps
//////////////////////////////////////////////////////////////////////////////////
// Company: SDSU
// Engineer: Colton Beery
// 
// Create Date: 02/20/2019 08:59:18 AM
// Revision Date: 3/6/2019 11:41 AM
// Module Name: UART_TX
// Project Name: UART
// Target Devices: Basys3
//
// Description: Tx: The simpler of the two. When an 8 bit value is loaded into a 
//              register using the 8 DIP switches for the number and a push button
//              for the "load" signal, it shifts the byte out in asynchronous 
//              serial format (initially at 9600 bits per second, later at an 
//              arbitrary, programmable data rate).  That begins with a start bit (0),
//              followed by the 8 data bits LSB first, and a stop (1) bit.
// 
// Dependencies: Basys3_Master_Customized.xdc
// 
// Revision History 
// Current Revision: 0.19
// Changelog in Changelog.txt
//
// Additional Comments:  
// 
//////////////////////////////////////////////////////////////////////////////////


module UART_TX(
    input [7:0] IO_SWITCH, //IO Dipswitches; up = 1
    input IO_BTN_C,         //IO Pushbutton (Center); pushed = 1
    input clk,              //Master clock signal
    output reg [7:0] JA,     //PMOD JA; port JA1 used as TX pin
    output wire [7:0] IO_LED     //IO LED's for debugging data
//    output wire [9:0] IO_LED     //IO LED's for debugging transmit
    );
    
    /* State Machine Parameters */
    reg [1:0] state = 0; 
    parameter idle = 2'b00; //State 0 = idle
    parameter start = 2'b01; //State 1 = start bit
    parameter out = 2'b10;  //State 2 = output
    parameter stop = 2'b11; //State 3 = stop bit
    
    /* Data and transmission parameters */
    reg [7:0] data = 0;         //data input    
//    reg [9:0] transmission; //what to output
    reg [3:0] bit = 0;                           //bit number currently being transmitted 
    
    parameter max_counter = 10415;          // this should give 9600 baud
    reg [13:0] counter = 0;                 //counter for baud rate generation; currently hardcoded to 14 bits for 9600 baud
    
    /* LED Debugging */
    assign IO_LED = data; //LEDs used to check if data is read successfully
//    assign IO_LED = transmission; //LEDs used to check if data is read successfully
    
    always @(posedge clk) begin
        
        /* Current State Logic */    
          case(state)
            idle: begin              
                JA[0] <= 1; //When idle, assert idle bit (1)
                /* Read Logic */
                if (IO_BTN_C) begin
                    data <= IO_SWITCH[7:0];     //read switches 1-8, where 0 is LSB
//                    transmission = {stop_bit, data, start_bit}; //transmission is Start->data (lsb first)->Stop
                    //IO_LED = transmission;
                    state <= start;
                    bit = 0;
                end                 
            end
            
            /* start bit */
            start: begin
//                for (counter = 0; counter < max_counter; counter = counter + 1) begin //if counter hasn't overflowed yet, transmit
//                    JA[0] <= 0; //start bit 0                        
//                end
                if (counter < max_counter) begin //if counter hasn't overflowed yet, transmit start bit
                        JA[0] <= 0;
                        counter <= counter + 1; 
                    end else begin //when counter overflows, done transmitting start bit, go to data
                        counter <= 0; 
                        state <= start;
                end
                state <= out;
            end
            
            /* Data transmission */
            out: begin
//                    for (bit = 0; bit <= 7; bit = bit + 1) begin // If there's still more bits to transmit
//                      for (counter = 0; counter < max_counter; counter = counter + 1) begin //if counter hasn't overflowed yet, transmit
//                            JA[0] <= data[bit]; 
//                      end
//                    end
                    if (bit <= 7) begin // If there's still more bits to transmit
                        if (counter < max_counter) begin //if counter hasn't overflowed yet, transmit
                            JA[0] <= data[bit];
                            counter <= counter + 1; 
                        end else begin //reset counter when it overflows, and go to next bit
                            counter <= 0;
                            bit <= bit + 1; 
                        end
                    end else begin //when you run out of bits to transmit, reset counter and bit
                        counter <= 0;
                        bit <= 0;  
                        state <= stop;
                    end                   
                end
            
            /* stop bit */
            stop: begin
//                for (counter = 0; counter <= max_counter; counter = counter + 1) begin //if counter hasn't overflowed yet, transmit
//                    JA[0] <= 1;  //stop bit is 1
//                end    
                if (counter < max_counter) begin //if counter hasn't overflowed yet, transmit
                        JA[0] <= 0;
                        counter <= counter + 1; 
                    end else begin //when counter overflows, done transmitting, go back to idle
                        counter <= 0; 
                        state <= idle;
                end
            end                                
        endcase    
    end
endmodule
